src/main/java/jenkins/plugins/slack/SlackUserIdResolver.java

```diff
@@ -0,0 +1,156 @@
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.util.EntityUtils;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+/**
+ *
+ * @author stuartr
+ */
+public class SlackUserIdResolver {
+
+    private static final Logger LOGGER = Logger.getLogger(SlackUserIdResolver.class.getName());
+
+    private static final String LOOKUP_BY_EMAIL_METHOD_URL_FORMAT = "https://slack.com/api/users.lookupByEmail?token=%s&email=%s";
+    private static final String SLACK_OK_FIELD = "ok";
+    private static final String SLACK_USER_FIELD = "user";
+    private static final String SLACK_ID_FIELD = "id";
+
+    private final String authToken;
+    private final CloseableHttpClient httpClient;
+    private final List<MailAddressResolver> mailAddressResolvers;
+
+    public SlackUserIdResolver(String authToken, CloseableHttpClient httpClient, List<MailAddressResolver> mailAddressResolvers) {
+        this.authToken = authToken;
+        this.httpClient = httpClient;
+        this.mailAddressResolvers = (mailAddressResolvers != null) ? mailAddressResolvers : MailAddressResolver.all();
+    }
+
+    public SlackUserIdResolver(String authToken, CloseableHttpClient httpClient) {
+        this(authToken, httpClient, null);
+    }
+
+    public List<String> resolveUserIdsForBuild(AbstractBuild build) {
+        return resolveUserIdsForChangeLogSets(build.getChangeSets());
+    }
+
+    public List<String> resolveUserIdsForRun(Run run) {
+        if (run instanceof RunWithSCM) {
+            RunWithSCM r = (RunWithSCM) run;
+            return resolveUserIdsForChangeLogSets(r.getChangeSets());
+        } else {
+            return Collections.EMPTY_LIST;
+        }
+    }
+
+    public List<String> resolveUserIdsForChangeLogSet(ChangeLogSet changeLogSet) {
+        return Arrays.stream(changeLogSet.getItems())
+                .map(item -> resolveUserId(((Entry) item).getAuthor()))
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+    }
+
+    public List<String> resolveUserIdsForChangeLogSets(List<ChangeLogSet> changeLogSets) {
+        return changeLogSets.stream()
+                .map(changeLogSet -> resolveUserIdsForChangeLogSet(changeLogSet))
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
+    }
+
+    public List<String> resolveUserIdsForEmailAddresses(List<String> emailAddresses) {
+        return emailAddresses.stream()
+                .map(address -> resolveUserIdForEmailAddress(address))
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+    }
+
+    public String resolveUserId(User user) {
+        for (MailAddressResolver resolver : mailAddressResolvers) {
+            String emailAddress = resolver.findMailAddressFor(user);
+            if (StringUtils.isNotEmpty(emailAddress)) {
+                String userId = resolveUserIdForEmailAddress(emailAddress);
```

Comment:
Is it worth caching this on the Jenkins user's user profile, so that you don't need to contact slack via api? i.e. it's unlikely that a user id would ever change and if it did a user could go in and clear it or update it. cc @casz

Reply:
I don't have experience adding data to a Jenkins user's profile. After some quick research I believe I need to create a new  type that extends [UserProperty](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/model/UserProperty.java)? I'll put something together after resolving your other suggestions.